include/dynd/kernels/string_algorithm_kernels.hpp

```diff
@@ -3,83 +3,33 @@
 // BSD 2-Clause License, see LICENSE.txt
 //
 
-// Implement a number of string algorithms. Note that the initial
+// Implement a number of string algorithms.
 
 #pragma once
 
+#include <dynd/callable.hpp>
 #include <dynd/kernels/kernel_builder.hpp>
-#include <dynd/string_encodings.hpp>
+#include <dynd/types/string_type.hpp>
 
-namespace dynd { namespace kernels {
+namespace dynd {
+  namespace nd {
 
-/**
- * String concatenation kernel.
- *
- * (string, string, ...) -> string
- */
-struct DYND_API string_concatenation_kernel {
-    typedef string_concatenation_kernel extra_type;
+    struct string_concatenation_kernel
+      : base_kernel<string_concatenation_kernel, 2> {
+      typedef string_concatenation_kernel extra_type;
 
-    ckernel_prefix m_base;
-    // The number of input operands
-    size_t m_nop;
+      void single(char *dst, char *const *src);
+    };
 
-    ckernel_prefix& base() {
-        return m_base;
-    }
+  } // namespace nd
 
-    /**
-     * Initializes the kernel data.
-     *
-     * \param nop  This must be >= 2.
-     * \param dst_arrmeta  Must be the destination for a "string" type (utf-8 string type).
-     * \param src_arrmeta  Must be the two sources for "string" types.
-     */
-    void init(size_t nop, const char *dst_arrmeta, const char **src_arrmeta);
+  namespace ndt {
 
-    static void single(char *dst, char *const *src,
-                       ckernel_prefix *extra);
-    static void strided(char *dst, intptr_t dst_stride, char *const *src,
-                        const intptr_t *src_stride, size_t count,
-                        ckernel_prefix *extra);
-};
+    template<>
+    struct traits<dynd::nd::string_concatenation_kernel> {
+      static type equivalent() { return callable_type::make(make_type<dynd::string>(), {type(string_id), type(string_id)}); }
+    };
 
```

Comment:
The `ndt::traits<...>` template that you are using communicates the signature of the kernel to how we make callables (details to follow on that).